src/Rules/Variables/CoalesceRule.php

```diff
@@ -0,0 +1,74 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Variables;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\NullType;
+use PHPStan\Type\VerbosityLevel;
+
+/**
+ * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Expr\BinaryOp\Coalesce>
+ */
+class CoalesceRule implements \PHPStan\Rules\Rule
+{
+
+	public function getNodeType(): string
+	{
+		return Node\Expr\BinaryOp\Coalesce::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$error = $this->canBeCoalesced($node->left, $scope);
+
+		if ($error === null) {
+			return [];
+		}
+
+		return [$error];
+	}
+
+	private function canBeCoalesced(Node $node, Scope $scope): ?RuleError
+	{
+		if ($node instanceof Node\Expr\Variable && is_string($node->name)) {
+
+			$hasVariable = $scope->hasVariableType($node->name);
+
+			if ($hasVariable->no()) {
+				return RuleErrorBuilder::message(
+					sprintf('Coalesce of undefined variable $%s.', $node->name)
+				)->line($node->getLine())->build();
+			}
+
+			$variableType = $scope->getVariableType($node->name);
+
+			if ($variableType->isSuperTypeOf(new NullType())->no()) {
+				return RuleErrorBuilder::message(
+					sprintf('Coalesce of variable $%s, which cannot be null.', $node->name)
+				)->line($node->getLine())->build();
+			}
+
+		} elseif ($node instanceof Node\Expr\ArrayDimFetch && $node->dim !== null) {
+			$type = $scope->getType($node->var);
+			$dimType = $scope->getType($node->dim);
+
+			if ($type->isOffsetAccessible()->no() || $type->hasOffsetValueType($dimType)->no()) {
+				return RuleErrorBuilder::message(
+					sprintf(
+						'Coalesce of invalid offset %s on %s.',
+						$dimType->describe(VerbosityLevel::value()),
+						$type->describe(VerbosityLevel::value())
+					)
+				)->line($node->getLine())->build();
+			}
+
+			return $this->canBeCoalesced($node->var, $scope);
```

Comment:
What about `$type->hasOffsetValueType($dimType)->yes()` and not null type? In that case the left side is also always used.

Reply:
Well, I figured that even though it might have the correct offset type, the offset might still not be set.

E.g.

```php
/** @var string */
$array = ['a' => 1];

echo $array['b'] ?? 0; // hasOffsetValueType = yes, type is not null, but the value is not set
```

Reply:
Ah wait, I misunderstood. I've handled this case now, and added a test.